schnorr_fun/src/frost/shared_key.rs

```diff
@@ -0,0 +1,234 @@
+    #[must_use]
+    pub fn homomorphic_add(
+        mut self,
+        tweak: Scalar<impl Secrecy, impl ZeroChoice>,
+    ) -> SharedKey<Normal, Zero> {
+        self.point_polynomial[0] = g!(self.point_polynomial[0] + tweak * G).normalize();
+        SharedKey::from_inner(self.point_polynomial)
+    }
+
+    /// Negates the polynomial
+    #[must_use]
+    pub fn homomorphic_negate(mut self) -> SharedKey<Normal, Z> {
+        poly::point::negate(&mut self.point_polynomial);
+        SharedKey::from_inner(self.point_polynomial)
+    }
+
+    /// Multiplies the shared key by a scalar.
+    ///
+    /// In order for a [`PairedSecretShare`] to be valid against the new key they will have to apply
+    /// [the same operation](super::PairedSecretShare::homomorphic_mul).
+    ///
+    /// [`PairedSecretShare`]: super::PairedSecretShare
+    #[must_use]
+    pub fn homomorphic_mul(mut self, tweak: Scalar<impl Secrecy>) -> SharedKey<Normal, Z> {
+        for coeff in &mut self.point_polynomial {
+            *coeff = g!(tweak * coeff.deref()).normalize();
+        }
+        SharedKey::from_inner(self.point_polynomial)
+    }
+
+    /// Create a shared key from a subset of verification shares.
+    ///
+    /// If all the verification shares are correct and you have at least a threshold of them then
+    /// you'll get the right answer. If you put in a wrong share you won't get the right answer!
+    ///
+    /// ## Security
+    ///
+    /// ⚠ You can't just take any random points you want and pass them in here and hope it's secure.
+    /// They need to be from a securely generated key.
+    pub fn from_verification_shares(
+        shares: &[(PartyIndex, Point<impl PointType, Public, impl ZeroChoice>)],
+    ) -> SharedKey<Normal, Z> {
+        let poly = poly::point::interpolate(shares);
+        let poly = poly::point::normalize(poly);
+        SharedKey::from_inner(poly.collect())
+    }
+}
+
+impl SharedKey<Normal> {
+    /// Convert the key into a BIP340 "x-only" SharedKey.
+    ///
+    /// This is the [BIP340] compatible version of the key which you can put in a segwitv1 output.
+    ///
+    /// [BIP340]: https://bips.xyz/340
+    pub fn into_xonly(mut self) -> SharedKey<EvenY> {
+        let needs_negation = !self.key().is_y_even();
+        if needs_negation {
+            self = self.homomorphic_negate();
+            debug_assert!(self.key().is_y_even());
+        }
+
+        SharedKey::from_inner(self.point_polynomial)
+    }
+}
+
+impl<Z: ZeroChoice> SharedKey<Normal, Z> {
+    /// The key that was shared with this polynomial defining the sharing.
+    ///
+    /// This is the first coefficient of the polynomial.
+    pub fn key(&self) -> Point<Normal, Public, Z> {
+        Z::cast_point(self.point_polynomial[0]).expect("invariant")
+    }
+    /// Constructor to create a from a vector of points where each item represent a polynomial
+    /// coefficient.
+    ///
+    /// Returns `None` if the first coefficient is [`Point::zero`].
```

Comment:
This is not true anymore. It can be `None` is the poly is empty. I think I might make two different `from_poly`s one for `Zero` and another for `NonZero`. The `Zero` one will be infallible.